### ISSUE TITLE
TypeScript silently omits enum assignment type error between two enums with the same name

### ISSUE THREAD

**Issue by evanw** [microsoft/TypeScript#55915]
### 🔎 Search Terms

I'm not exactly sure what keywords to use for this because it's kind of a strange problem. Some ideas: enum assignment same name error missing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (both TypeScript 3.3 and TypeScript 5.2)

### 💻 Code

I couldn't do a playground link for this because it requires two separate files:

* `types.ts`

    ```ts
    export enum DiagnosticCategory {
        Warning,
        Error,
        Suggestion,
        Message,
    }

    export enum DiagnosticCategory2 {
        Warning,
        Error,
        Suggestion,
        Message,
    }
    ```

* `diagnosticInformationMap.generated.ts`

    ```ts
    import { DiagnosticCategory as DiagnosticCategory2 } from "./types";
    // import { DiagnosticCategory2 as DiagnosticCategory2 } from "./types";

    enum DiagnosticCategory {
      Warning = 'Warning',
      Error = 'Error',
      Suggestion = 'Suggestion',
      Message = 'Message',
    }

    function diag2(category1: DiagnosticCategory, category2: DiagnosticCategory2) {
      category1 = category2;
      category2 = category1;
    }
    ```

### 🙁 Actual behavior

There is no error if you import `DiagnosticCategory as DiagnosticCategory2` but there is an error if you `DiagnosticCategory2 as DiagnosticCategory2`. My `tsconfig.json` for this experiment:

```json
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "strict": true
  }
}
```

Demo:

```console
$ node_modules/.bin/tsc -version
Version 5.2.2

$ head -n 2 diagnosticInformationMap.generated.ts 
import { DiagnosticCategory as DiagnosticCategory2 } from "./types";
// import { DiagnosticCategory2 as DiagnosticCategory2 } from "./types";

$ node_modules/.bin/tsc -p .
(no errors)

$ head -n 2 diagnosticInformationMap.generated.ts 
// import { DiagnosticCategory as DiagnosticCategory2 } from "./types";
import { DiagnosticCategory2 as DiagnosticCategory2 } from "./types";

$ node_modules/.bin/tsc -p .
diagnosticInformationMap.generated.ts:12:3 - error TS2322: Type 'DiagnosticCategory2' is not assignable to type 'DiagnosticCategory'.

12   category1 = category2;
     ~~~~~~~~~

diagnosticInformationMap.generated.ts:13:3 - error TS2322: Type 'DiagnosticCategory' is not assignable to type 'DiagnosticCategory2'.

13   category2 = category1;
     ~~~~~~~~~


Found 2 errors in the same file, starting at: diagnosticInformationMap.generated.ts:12
```

### 🙂 Expected behavior

I expected importing `DiagnosticCategory as DiagnosticCategory2` to cause the same error that importing `DiagnosticCategory2 as DiagnosticCategory2` does. My rationale is that the two type definitions in `types.ts` appear to be identical to me, so my intuition tells me to expect them to behave identically. It seems to me like TypeScript is perhaps silently swallowing the error about assignment compatibility if the two types are named the same thing even though the names refer to two separate incompatible things from two separate files.

### Additional information about the issue

I discovered this confusing behavior while helping @jakebailey on #55894. It seemed like a bug to me so I'm reporting it here. Feel free to close this issue if this is expected behavior.

**Comment by DanielRosenwasser** [comment]
What an enormous WAT.

[Here is a single-file example](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwCIEsCGBzRCAZxiwGMBhHGYPaATzgG8BYAKDk7gHUcpEsiPABp2XOAFEoUaKI5cAysjx5gJLBERzxAWTVF8wOQF92iHKjVgcZYHBzN2YrqEiwEKdNnyF1larQMAEyO8uKcvPyCIs7hktKyseFKKmqkmtpxcHpEBqqZnKZsRWYWVjZ2AEah4q7Q8EhomLgExKT+NHRQjKxh4ZECQnAAvHAARAPRYwXhUjJQI+Nz0NNJ4imq6pqLYxtpGoirfbr6hjs5ecBH4kUlbABmyIhk6Yhw9wAUIABc9gB03lafionQYwjg9F+lQBLV87RBgW6AEoalwQIt6ABuJKMUYgbHFJwPJ4vA5wPBfX44GE+NrkBFdehBcGQuDQwFw+kBRko3ridGjLE4xb49hFIA):

```ts
namespace a {
    export enum DiagnosticCategory {
        Warning,
        Error,
        Suggestion,
        Message,
    }

    export enum DiagnosticCategory2 {
        Warning,
        Error,
        Suggestion,
        Message,
    }
}

namespace b {
    export enum DiagnosticCategory {
        Warning = "Warning",
        Error = "Error",
        Suggestion = "Suggestion",
        Message = "Message",
    }
}

function f(x: a.DiagnosticCategory, y: b.DiagnosticCategory) {
    x = y;
    y = x;
}

function g(x: a.DiagnosticCategory2, y: b.DiagnosticCategory) {
    x = y;
    y = x;
}
```

and [here is one that doesn't use `namespace` declarations at all](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwCIEsCGBzRCAZxiwGMBhHGYPaATzgG8BYAKDk7gHUcpEsiPABp2XOAFEoUaKI5cAysjx5gJLBERzxAWTVF8wOQF92oSLDgAbYPBAAuTLgLFSlarQbt2reZyRoTviE6u40dFCMvuLivPyCeHAAvHAARHECQqnaMZxSMlDJafnQ2WK5cEoqaqSaRalVquqaZX4xekQGqvUdXcCt4qZtAGbIiGS1iHDDABT0jtjBruRU4QwAlMzluSBF9ADc2zGMKSCHbUNDQA):

```ts
export enum DiagnosticCategory {
    Warning,
    Error,
    Suggestion,
    Message,
}
export let x: DiagnosticCategory

{
    enum DiagnosticCategory {
        Warning = "Warning",
        Error = "Error",
        Suggestion = "Suggestion",
        Message = "Message",
    }
    function f(y: DiagnosticCategory) {
        x = y;
        y = x;
    }
}
```

So the context is that ever since TypeScript 1.8(?), a source enum is are compatible with a target enum if its names (the enum declaration and its members) are all identical.

- https://github.com/microsoft/TypeScript/issues/1748
- https://github.com/Microsoft/TypeScript/issues/5740
- https://github.com/microsoft/TypeScript/issues/5943
- https://github.com/microsoft/TypeScript/pull/6036

I think the weird thing here is that we never checked whether the enum values even though that was explicitly mentioned in the design notes at #5943. I don't know if @sandersn knows why that is though.

Part of the issue I think was that people got frustrated over mismatches in numeric enum values. Adding a new member can shift an entire range of enums.

Still, I don't know if I find that all that convincing. Ignoring the values seems extremely wrong.

**Comment by fatcerberus** [comment]
> a source enum is are compatible with a target enum if its names (the enum declaration and its members) are all identical

tl;dr someone took the term "nominal typing" too literally
